### ISSUE TITLE
Audio buffers improperly without clear cause

### ISSUE THREAD

**Issue by Ophiolith** [fgsfdsfgs/perfect_dark#222]
At completely arbitrary points during gameplay, Perfect Dark's audio buffer timing seems to get thrown out, causing 'stretched', stuttery audio that lags significantly behind cues, which persists until the game executable is closed. This occurs regardless of whether vsync is turned on or off, or whether a frame limit is set, however the problem seems to be exacerbated when vsync is disabled. Adjusting the audio buffer size or queue limit does not seem to affect it, as I have tried on much larger and smaller values for both settings without any noticeable effect.
It seems to occur most commonly when the game exhibits minor frame-time variances, such as on more complex levels like Carrington Villa, or when framebuffer effects are triggered, but not always. My PC is not brand new, but plays the game at an almost entirely locked 60fps with very minimal spikes.

**Comment by fgsfdsfgs** [comment]
I've never had this happen. Are you on Linux? Which region/ROM are you using?

**Comment by Ophiolith** [comment]
Nope, on Windows 10 22H2 with a verified good ntsc-final 1.1 rom. The only other thing possibly worth mentioning is that I'm using a Focusrite Scarlett 6i6 external audio interface running at a samplerate of 48khz by default. I'm not sure how much relevance that might bear.

**Comment by fgsfdsfgs** [comment]
It might have some relevance, but I don't know for sure. The game audio is output at a somewhat non-standard sample rate of 22020. Maybe it would be a better idea to have an option to resample it to 22050 or a user specified sample rate. I don't know if that would cause the problem you're describing though.

**Comment by Ophiolith** [comment]
I've just had another run-through on Villa after changing the interface's samplerate to 44.1khz, and this seems to have improved things. I'll run a few more tests tomorrow and see if I can reproduce it, but I accept it may just be a bit of a corner case with my particular setup.

**Comment by fgsfdsfgs** [comment]
Even if it is that, I think I should still add an option to the port to avoid this.
To be fair, the whole audio output syste is currently a bit wack.

**Comment by Ophiolith** [comment]
I'm definitely not well versed enough to know how SDL handles audio internally, but AFAIK most audio interfaces don't natively support samplerates that low, so it could be possible there's already resampling going on at the driver level? The interface I'm using tends to set the driver-specified samplerate pretty aggressively unless it's told otherwise, so there could be something going on there that's throwing things off?

**Comment by Ophiolith** [comment]
Changing the samplerate in the driver settings was not a reliable fix. The issue still reoccurs. Curiously, I've noticed that the issue gradually clears up over the course of a minute or two, becoming less severe before becoming contiguous again, however the issue will continue to reoccur even after this.
~~However, for whatever reason, forcing on vsync and forcing triple buffering via the Nvidia control panel seems to have drastically improved things. I'm thinking now that something in the audio code is extremely fussy about frame timing? Will continue to test.~~
EDIT: Forcing vsync does not resolve either - it's like sometimes the game will boot up and run absolutely fine and no audio issues occur for the duration of play. Other times the audio is just consistently broken. I cannot understand what's causing this at all.

**Comment by captinwill** [comment]
If it helps i am currently having this issue, specs are
Windows 10 22H2
Ryzen 5 5600x 6 Core 12 Thread
RTX 3060 12 GB
Using most up to date version of port.
